### DISCUSSION TITLE
Function Docstyle

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#878] [General] [answered]
Whilst looking at https://github.com/home-assistant/core/pull/89035 I realised that ruff has a `convention` option for pydocstyle:
https://beta.ruff.rs/docs/faq/#does-ruff-support-numpy-or-google-style-docstrings
```toml
[tool.ruff.pydocstyle]
convention = "google/numpy/pep257"
```

A list of rules for each is also available on this commit https://github.com/charliermarsh/ruff/commit/ca58c72fc9349547eff753acd5aa2818cadbd6b1

This is based on https://www.pydocstyle.org/en/stable/error_codes.html#default-conventions

I think that we should make a decision as to which style we recommend going forward.

See https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats

Developer documentation has been updated: https://github.com/home-assistant/developers.home-assistant/pull/2797

**Comment by epenet** [comment]
Proposal 1: `numpy` uses dashed underline for sections
Disables: D107, D203, D212, D213, D402, D413, D415, D416, and D417
https://numpydoc.readthedocs.io/en/latest/format.html#parameters
```python
"""
My numpydoc description of a kind
of very exhautive numpydoc format docstring.

Parameters
----------
first : array_like
    the 1st param name `first`
second :
    the 2nd param
third : {'value', 'other'}, optional
    the 3rd param, by default 'value'

Returns
-------
string
    a value in a string

Raises
------
KeyError
    when a key error
OtherError
    when an other error
"""
```

**Reply by emontnemery** [reply]
This style is supported by vscode:
![image](https://github.com/user-attachments/assets/5a18ae97-a2da-40df-bcbc-12529f47dff2)

**Comment by epenet** [comment]
Proposal 2: `google`
Disables: D203, D204, D213, D215, D400, D401, D404, D406, D407, D408, D409 and D413
https://google.github.io/styleguide/pyguide.html#383-functions-and-methods
```python
"""
This is an example of Google style.

Args:
    param1: This is the first param.
    param2: This is a second param.

Returns:
    This is a description of what is returned.

Raises:
    KeyError: Raises an exception.
"""
```

**Reply by emontnemery** [reply]
This style is supported by vscode:
![image](https://github.com/user-attachments/assets/7cb97a44-4498-40e1-b845-9e5fa9048c5d)

**Comment by epenet** [comment]
Proposal 3: `pep257`
Disables: D203, D212, D213, D214, D215, D404, D405, D406, D407, D408, D409, D410, D411, D413, D415, D416 and D417

**Comment by epenet** [comment]
Proposal 4: we can also keep `manual` configuration of the individual ignore flags
Currently disabled: D202/D203/D213/D404/D406/D407/D411

In this scenario, I would suggest we only ignore D203 (which is incompatible with D211) and D213 (which is incompatible with D212)
And we should enable D202, D404, D411
I am not 100% sure about D406, D407

**Comment by emontnemery** [comment]
We are using [sphinx style docstrings](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html) in some cases, vscode supports this style.

```py
    @handle_boto_errors
    async def async_download_backup(
        self,
        backup_id: str,
        **kwargs: Any,
    ) -> AsyncIterator[bytes]:
        """Download a backup file.

        :param backup_id: The ID of the backup that was returned in async_list_backups.
        :return: An async iterator that yields bytes.
        """
```

![image](https://github.com/user-attachments/assets/47cfc2c7-e533-4374-8126-caf762dbcd04)

**Reply by epenet** [reply]
Ruff does not yet support this style: https://github.com/astral-sh/ruff/pull/13286

**Comment by emontnemery** [comment, marked as answer]
The proposal was approved in the core meeting July 3rd:
- The decision is to go for the Google style as it is easy to read both for humans and computers (ruff has rules, and vscode supports it)
- Docstrings which document parameters, return value and exceptions should follow the Google style
- We don't currently see a need to globally enforce docstrings to follow this pattern; thanks to modern type annotations the function signature is often enough
- Something which is always welcome is add a "Raises" section to the docstring
- Focus should be to add improved docstrings to core parts where google-style docstrings add value, not integrations

**Reply by epenet** [reply]
Documentation added in https://github.com/home-assistant/developers.home-assistant/pull/2797

**Comment by emontnemery** [comment]
Developer documentation has been updated: https://github.com/home-assistant/developers.home-assistant/pull/2797
